skfem/assembly/basis/facet_basis.py

```diff
@@ -108,10 +115,110 @@ def default_parameters(self):
+                raise NotImplementedError
+            # special case: piecewise constant elem
+            I = fbasis.dofs.interior_dofs[:, self.tind].flatten()
+        y = project(x, self, fbasis, I=I)
+
+        # build connectivity for lower dimensional mesh
+        ix = self.mesh.facets[:, self.find]
+        ixuniq = np.unique(ix)
+        b = np.zeros(np.max(ix) + 1, dtype=np.int64)
+        b[ixuniq] = np.arange(len(ixuniq), dtype=np.int64)
+
+        return self.mesh.p[:, ixuniq], b[ix], y
+
+    def trace(self,
+              x: ndarray,
+              projection: Callable[[ndarray], ndarray],
+              target_elem: Optional[Element] = None) -> Tuple[InteriorBasis,
+                                                              ndarray]:
+        """Restrict solution to :math:`d-1` dimensional trace mesh.
+
+        The user must define how the boundary points are projected using the
+        argument ``projection``.  For example,
+
+        >>> projection = lambda p: p[0]
+
+        will keep only the `x`-coordinate when initializing the trace mesh.
+
+        Parameters
+        ----------
+        x
+            The solution vector.
+        projection
+            A function defining the projection of the boundary points.  See
+            above for an example.
+        target_elem
+            Optional lower-dimensional finite element to project to.  If not
+            given, a piecewise constant element is used.
+
+        Returns
+        -------
+        InteriorBasis
+            An object corresponding to the trace mesh.
+        ndarray
+            A projected solution vector defined on the trace mesh.
+
+        """
+        DEFAULT_TARGET = {
+            MeshTri: ElementLineP0,
+            MeshQuad: ElementLineP0,
+            MeshTet: ElementTriP0,
+            MeshHex: ElementQuad0,
+        }
+
+        meshcls = type(self.mesh)
+        if meshcls not in DEFAULT_TARGET:
+            raise NotImplementedError("Mesh type not supported.")
+        if target_elem is None:
+            target_elem = DEFAULT_TARGET[meshcls]()
+
+        TRACE_RESTRICT_MAP: Dict[Tuple[Type[Element], Type[Mesh]], Any] = {
+            (ElementLineP0, MeshTri): (ElementTriP0, MeshLine),
+            (ElementLineP1, MeshTri): (ElementTriP1, MeshLine),
+            (ElementLineP2, MeshTri): (ElementTriP2, MeshLine),
+            (ElementLineP0, MeshQuad): (ElementQuad0, MeshLine),
+            (ElementLineP1, MeshQuad): (ElementQuad1, MeshLine),
+            (ElementLineP2, MeshQuad): (ElementQuad2, MeshLine),
+            (ElementTriP0, MeshTet): (ElementTetP0, MeshTri),
+            (ElementTriP1, MeshTet): (ElementTetP1, MeshTri),
+            (ElementTriP2, MeshTet): (ElementTetP2, MeshTri),
+            (ElementQuad0, MeshHex): (ElementHex0, MeshQuad),
+            (ElementQuad1, MeshHex): (ElementHex1, MeshQuad),
+            (ElementQuad2, MeshHex): (ElementHex2, MeshQuad),
+        }
+
```

Comment:
Is the second term of the value redundant?
```python
all(k[0].mesh_type is v[1] for k, v in TRACE_RESTRICT_MAP.items())
```

Reply:
Sure, thanks. I'll fix it.